Fix leftnav icon tooltip and showing user info in side leftnav nav.

kavyanekkalapu opened the pull request:
In this PR, Fixed few issues.

- Added icon tooltip for menu list when menu is collapsed
- Added tooltip for search and group name when menu is collapsed
- Added tabindex for search so that search can be focused.
- Fixed bug about not showing userinfo and grouinfo (those two were still in collapsed state)when user tab into leftnav.
- Fixed showing user name info when user tab into side nav.

**Before:**
No tooltips for icons.

<img width="227" alt="Screen Shot 2020-10-11 at 5 27 49 PM" src="https://user-images.githubusercontent.com/37085529/95690627-1e61aa00-0be7-11eb-866b-4a3bc6830671.png">

**After:**
<img width="293" alt="Screen Shot 2020-10-12 at 10 35 07 PM" src="https://user-images.githubusercontent.com/37085529/95808788-3d416880-0cdb-11eb-899c-190ece0ebd99.png">

<img width="548" alt="Screen Shot 2020-10-12 at 10 35 13 PM" src="https://user-images.githubusercontent.com/37085529/95808800-429eb300-0cdb-11eb-9826-551152f982d7.png">


fyi @skateman @himdel 


Parent issue: https://github.com/ManageIQ/manageiq-ui-classic/issues/7358

skateman commented on app/javascript/menu/side-nav-menu-link.jsx:
```diff
@@ -27,7 +29,15 @@ const SideNavMenuLink = ({
         {IconElement && (
           <SideNavIcon small>
-            <IconElement />
+            {expanded && (<IconElement />)}
```
I don't think we need to overcomplicate this, what about having the tooltip even if the menu is not collapsed?

kavyanekkalapu replied:
There are two issues by doing that.

- Screen reader will reader both tooltip and text beside it, then it will be like example: Network Network , which seems confusing experience to users 
- when we tab into left nav, tab focus go to icon first as it has tooltip  and then anchor tag, we tab through leftnav for navigation, not to read tooltips, when text is already present, tooltip adding with same text feels completely redundant annoying experience for keyboard users.

<img width="349" alt="Screen Shot 2020-10-12 at 9 10 30 AM" src="https://user-images.githubusercontent.com/37085529/95750647-7215d700-0c6b-11eb-91f5-f605ecb8fcbf.png">

skateman commented on app/javascript/menu/main-menu.jsx:
```diff
@@ -53,6 +54,7 @@ export const MainMenu = ({
   useEffect(() => {
     // cypress, debugging
     window.ManageIQ.menu = menu;
+    console.log("event.target-----------",document.activeElement);
```
You probably don't want this in your final code...

kavyanekkalapu replied:
yes @skateman i added a line in description that there is a bug i will work on it and will remove console logs then.

kavyanekkalapu replied:
Removed the console log

skateman commented on app/javascript/menu/main-menu.jsx:
```diff
@@ -61,13 +63,27 @@ export const MainMenu = ({
+    return () => {
+      window.removeEventListener('keydown', toggleExpanded);
+    };
+  });
```
Umm, what does this do? Why does it have a `console.log`?

kavyanekkalapu replied:
yes @skateman i added a line in description that there is a bug i will work on it and will remove console logs then.

skateman replied:
Sorry, I have a tendency of reading code first :see_no_evil: :disappointed_relieved:

kavyanekkalapu replied:
Remove the console.log

kavyanekkalapu commented:
@skateman @himdel  PR is ready for review, can you please review?

himdel commented on app/stylesheet/menu.scss:
```diff
@@ -123,3 +123,41 @@
+    align-items: center;
+  }
+
+  .vertical-center {
```
also, `.vertical-center` is already defined on line 70

kavyanekkalapu replied:
Thanks Fixed it, please check the updated code

kavyanekkalapu commented on app/stylesheet/menu.scss:
```diff
@@ -123,3 +123,41 @@
+  }
+}
+
+#main-menu {
```
Thanks Fixed it, please check the updated code

skateman commented on app/stylesheet/menu.scss:
```diff
@@ -123,3 +140,18 @@
   margin: -4px 0 0 -8px;
 }
+
+.bx--side-nav__item {
```
Is this living outside the menu? If not, can you move this inside the `#main-menu` to the same level as `.bx--side-nav.secondary`?

kavyanekkalapu replied:
Fixed it. please check updated code.

kavyanekkalapu commented on app/javascript/menu/main-menu.jsx:
```diff
@@ -66,20 +67,35 @@ export const MainMenu = ({
   }, []);
 
+  useEffect(() => {
+    function toggleExpanded(event) {
```
Changed the call. please check updated code.

kavyanekkalapu commented on app/javascript/menu/main-menu.jsx:
```diff
@@ -66,20 +67,35 @@ export const MainMenu = ({
 
+  useEffect(() => {
+    function toggleExpanded(event) {
+      if (event.key === 'Tab' && sidenavContainer.current && sidenavContainer.current.contains(event.target) && !event.target.classList.contains('menu-collapse-button')) {
```
Fixed it. please check updated code.

kavyanekkalapu commented on app/javascript/menu/main-menu.jsx:
```diff
@@ -66,20 +67,35 @@ export const MainMenu = ({
+    }
+    window.addEventListener('keydown', toggleExpanded);
+    window.addEventListener('focusin', toggleExpanded);
+    return () => {
```
Fixed it. please check updated code.

kavyanekkalapu commented on app/javascript/menu/main-menu.jsx:
```diff
@@ -66,20 +67,35 @@ export const MainMenu = ({
+        currentUser={currentUser}
+        brandUrl={brandUrl}
+      />
+      <div onClick={hideSecondary} KeyDown={hideSecondaryEscape} role="navigation" id="main-menu-primary" ref={sidenavContainer}>
```
Fixed it. please check updated code.

skateman commented:
After this change focusing on the menu in collapsed using the keyboard automatically expands it. I'm not sure if this is desired :confused:

skateman commented:
Yes, but after unfocusing the menu, it collapses back again. After your change it doesn't ...

skateman commented on app/stylesheet/menu.scss:
```diff
@@ -118,6 +118,10 @@
       }
 
+      .pficon {
+        color: $field-01;
```
What's this doing?

kavyanekkalapu replied:
After doing bin/update, there is one more menu item added in my lefnav menu, it is a pficon , not an svg, it is in blue color instead of above color. so i am overriding it so that that icon seems consistent with other svgs.

skateman commented on app/javascript/menu/main-menu.jsx:
```diff
@@ -69,20 +70,25 @@ export const MainMenu = ({
+            setExpanded(true);
+            setOpen(true);
+          }
+        } else if (openMenu && !activeSection) {
```
The if-if-if-else-if here is a little too much, can you please simplify it?

skateman commented:
After your changes, there's no way to access the 2nd level of the menu which is probably fine as selecting a 1st level item should eventually have the focus on the 2nd level's first item.

kavyanekkalapu commented:
> After your changes, there's no way to access the 2nd level of the menu which is probably fine as selecting a 1st level item should eventually have the focus on the 2nd level's first item.

i believe this change got introduced in https://github.com/ManageIQ/manageiq-ui-classic/pull/7406, it is in master too.

skateman commented:
It's still fine, my next PR will fix it.

kavyanekkalapu commented:
@skateman Removed ref and useEffect, added event listeners on div, also fixed some  eslint a11y issues based on the solution in https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-noninteractive-element-interactions.md.

skateman commented:
Clicking on a menu item while collapsed permanently expands the menu.

kavyanekkalapu commented:
> Clicking on a menu item while collapsed permanently expands the menu.

@skateman Fixed it , please check now, thanks.

skateman commented:
Your changes flip the expand/collapse chevron in the bottom part when focusing on the menu, which is confusing. When the menu is collapsed and you peek-expand it by focusing, the chevron should not change its direction.
**Before:**
![Screenshot from 2020-10-20 09-36-48](https://user-images.githubusercontent.com/649130/96555478-204afd80-12b8-11eb-9cb6-e8613c26a952.png)

**After:**
![Screenshot from 2020-10-20 09-38-45](https://user-images.githubusercontent.com/649130/96555498-28a33880-12b8-11eb-91b1-561ec11a46f5.png)

kavyanekkalapu commented:
> Your changes flip the expand/collapse chevron in the bottom part when focusing on the menu, which is confusing. When the menu is collapsed and you peek-expand it by focusing, the chevron should not change its direction.
> **Before:**
> ![Screenshot from 2020-10-20 09-36-48](https://user-images.githubusercontent.com/649130/96555478-204afd80-12b8-11eb-9cb6-e8613c26a952.png)
> 
> **After:**
> ![Screenshot from 2020-10-20 09-38-45](https://user-images.githubusercontent.com/649130/96555498-28a33880-12b8-11eb-91b1-561ec11a46f5.png)

@skateman I did not understand why this is an issue, as soon as we tab into leftnav, it is showing expanded chevron, as soon as we are out of leftnav, it is showing collapsed chevron.  For me functionality in master is really confusing, here on tabbing, leftnav is opening and chevron still showing as collapsed..then lets say i click on chevron it is opening again ( which is already opened before)

skateman commented:
Because as soon as you unfocus the menu, it jumps back...

kavyanekkalapu commented:
> Because as soon as you unfocus the menu, it jumps back...

Yes as soon as we unfocus leftnav, menu is closing and chevron icon came to previous sate which is collapsed state..

skateman commented:
Now the chevron is not focusable in collapsed mode, the first item to get focused on that triggers the peek is the Logout menu item. I suggest you to extract the tooltip fix into a separate PR as that part is okay and can be merged separately. This one might need some kind of automated tests that verify the correctness, because after every change I was able to find issues...

kavyanekkalapu commented:
> Now the chevron is not focusable in collapsed mode, the first item to get focused on that triggers the peek is the Logout menu item. I suggest you to extract the tooltip fix into a separate PR as that part is okay and can be merged separately. This one might need some kind of automated tests that verify the correctness, because after every change I was able to find issues...

@skateman Chevron is focusable in collapsed mode for me, are you trying in specific use case?

<img width="265" alt="Screen Shot 2020-10-23 at 9 13 45 AM" src="https://user-images.githubusercontent.com/37085529/97008117-6c907a80-1510-11eb-975a-e2caef808b6c.png">

skateman reviewed (approved):
![The Seal of Approval](http://www.skateman.eu/seal.jpg)
